pkg/hnsw/heap.go

```diff
@@ -0,0 +1,115 @@
+package hnsw
+
+import (
+	"fmt"
+)
+
+type Item struct {
+	id   Id
+	dist float32
+}
+
+var EmptyHeapError = fmt.Errorf("Empty Heap")
+
+type HeapInterface interface {
+	PeekMinItem() (*Item, error)
+	PeekMaxItem() (*Item, error)
+	PopMinItem() (*Item, error)
+	PopMaxItem() (*Item, error)
+	Insert(id Id, dist float32)
+	IsEmpty() bool
+	Len() int
+	Less(i, j int) bool
+	Swap(i, j int)
+	Push(x *Item)
+	Pop() *Item
+}
+
+type DistHeap struct {
+	items   []*Item
+	visited map[Id]bool
+}
+
+func NewDistHeap() *DistHeap {
+	d := &DistHeap{
+		items:   make([]*Item, 0),
+		visited: make(map[Id]bool),
+	}
+	Init(d)
+	return d
+}
+func FromItems(items []*Item) *DistHeap {
+	d := &DistHeap{items: items, visited: make(map[Id]bool)}
+	Init(d)
+	return d
+}
+func (d *DistHeap) PeekMinItem() (*Item, error) {
+	if d.IsEmpty() {
+		return nil, EmptyHeapError
+	}
+
+	return (*d).items[0], nil
+}
+func (d *DistHeap) PeekMaxItem() (*Item, error) {
+	if d.Len() == 0 {
+		return nil, EmptyHeapError
+	}
+
+	// Find the maximum element without removing it
+	n := d.Len()
+
+	i := 0
+	l := lchild(0)
+	if l < n && !d.Less(l, i) {
+		i = l
+	}
+
+	r := rchild(0)
+	if r < n && !d.Less(r, i) {
+		i = r
+	}
+
+	return (*d).items[i], nil
+}
+func (d *DistHeap) PopMinItem() (*Item, error) {
+	if d.IsEmpty() {
+		return nil, EmptyHeapError
+	}
+
+	return Pop(d).(*Item), nil
```

Comment:
This still seems to use the old interface, should this be `d.Pop()` instead?

Reply:
Whoops. Done